File: src/main/java/grokkingfp/ch02.java
package grokkingfp;

import java.util.ArrayList;
import java.util.List;

public class ch02 {
    public class ShoppingCart {
        private List<String> items = new ArrayList<>();
        // private boolean bookAdded = false - removing the state

        public void addItem(String item) {
            items.add(item);
        }

        public int getDiscountPercentage() {
            if (items.contains("Book")) {
                return 5;
            } else {
                return 0;
            }
        }

        public List<String> getItems() {
            return new ArrayList<>(items);
        }

        public void removeItem(String item) {
            items.remove(item);
        }
    }


    public void main(String[] args) {
        ShoppingCart cart = new ShoppingCart();
        cart.addItem("Apple");
        System.out.println(cart.getDiscountPercentage());

        cart.addItem("Book");
        System.out.println(cart.getDiscountPercentage());

    }
}

File: src/main/kotlin/grokkingfp/ch02.kt
package grokkingfp

class ch02 {
    class ShoppingCart(
        private val items: List<String> = emptyList()
    ) {

        // private boolean bookAdded = false - removing the state
        fun addItem(items: List<String>, item: String): ShoppingCart {
            items.plus(item)
            return ShoppingCart(items)
        }

        fun getItems(): List<String> = items

        fun removeItem(items: List<String>, item: String): List<String> = items.minus(item)

        companion object {
            fun getDiscountPercentage(items: List<String>): Int =
                if (items.contains("Book")) {
                    5
                } else {
                    0
                }
        }
    }


}

fun main(args: Array<String>) {
    println(ch02.ShoppingCart.getDiscountPercentage(listOf("Apple")))

    println(ch02.ShoppingCart.getDiscountPercentage(listOf("Apple", "Book")))

    val cart = ch02.ShoppingCart(listOf("Apple", "Book"))
    val updatedCart = cart.addItem(listOf("Apple", "Book"), "Banana")

    println(updatedCart.getItems())
}